pkg/pgclient/config.go

```diff
@@ -135,70 +139,65 @@ func (cfg *Config) GetConnectionStr() string {
-	maxProcs := runtime.GOMAXPROCS(-1)
-	if cfg.WriteConnections < 0 {
-		return 0, 0, 0, fmt.Errorf("invalid number of writer connections %v, cannot be a negative number", cfg.WriteConnections)
+// GetPoolSize returns the max pool size based on the max_connections allowed by database and the defaultFraction.
+// Arg inputPoolSize is the pool size provided in the CLI flag.
+func (cfg *Config) GetPoolSize(poolName string, defaultFraction float64, inputPoolSize int) (int, error) {
+	maxConns, err := cfg.maxConn()
+	if err != nil {
+		return 0, fmt.Errorf("max connections: %w", err)
 	}
 
+	if inputPoolSize == defaultPoolSize {
+		// For the default case, we need to take up defaultFraction of allowed connections.
+		poolSize := float64(maxConns) * defaultFraction
+		if cfg.UsesHA {
+			poolSize /= 2
+		}
+		return int(poolSize), nil
+	}
+
+	// Custom pool size,
+	if inputPoolSize < MinPoolSize {
+		return 0, fmt.Errorf("%s pool size canot be less than %d: received %d", poolName, MinPoolSize, inputPoolSize)
+	} else if inputPoolSize > maxConns {
+		return 0, fmt.Errorf("%s pool size canot be greater than the 'max_connections' allowed by the database", poolName)
+	}
+	if cfg.UsesHA {
+		inputPoolSize /= 2
+	}
+	return inputPoolSize, nil
+}
+
+func (cfg *Config) maxConn() (int, error) {
 	conn, err := pgx.Connect(context.Background(), cfg.GetConnectionStr())
 	if err != nil {
-		return 0, 0, 0, err
+		return 0, err
 	}
 	defer func() { _ = conn.Close(context.Background()) }()
 
-	max = cfg.MaxConnections
-	if max < 1 {
-		var maxStr string
-		row := conn.QueryRow(context.Background(), "SHOW max_connections")
-		err = row.Scan(&maxStr)
-		if err != nil {
-			return 0, 0, 0, err
-		}
-		max, err = strconv.Atoi(maxStr)
-		if err != nil {
-			log.Warn("err", err, "msg", "invalid value from postgres max_connections")
-			max = 100
-		}
-
-		//In HA setups
-		if cfg.UsesHA {
-			max = max / 2
-		}
+	var maxConns int
+	err = conn.QueryRow(context.Background(), "SELECT current_setting('max_connections')::int").Scan(&maxConns)
+	if err != nil {
+		return 0, fmt.Errorf("error getting 'max_connections': %w", err)
+	}
+	return maxConns, nil
+}
 
-		if max <= 1 {
-			log.Warn("msg", "database can only handle 1 connection")
-			return 1, 1, 1, nil
-		}
-		// we try to only use 80% the database connections, capped at 100
-		max = int(0.8 * float32(max))
-		if max > 100 {
-			max = 100
-		}
+func (cfg *Config) GetNumCopiers() (int, error) {
+	conn, err := pgx.Connect(context.Background(), cfg.GetConnectionStr())
+	if err != nil {
+		return 0, err
 	}
+	defer func() { _ = conn.Close(context.Background()) }()
 
+	numCopiers := 0
 	err = conn.QueryRow(context.Background(), "SELECT _prom_ext.num_cpus()").Scan(&numCopiers)
 	if err != nil {
-		return 0, 0, 0, fmt.Errorf("error fetching number of CPUs from extension: %w", err)
+		return 0, fmt.Errorf("error fetching number of CPUs from extension: %w", err)
 	}
 
 	if cfg.WriteConnections > 0 {
```

Comment:
I would do this first and just return the cfg value. No need to connect to the DB and fetch the number in this case.